lib/src/_shared/PickerPopper.tsx

```diff
@@ -0,0 +1,143 @@
+import { executeInTheNextEventLoopTick } from '../_helpers/utils';
+import { TransitionProps } from '@material-ui/core/transitions/transition';
+
+export interface ExportedPickerPopperProps {
+  /**
+   * Popper props passed down to material-ui [Popper](https://material-ui.com/api/popper/#popper-api) component.
+   */
+  PopperProps?: Partial<PopperProps>;
+  /**
+   * Custom component for [Transition](https://material-ui.com/components/transitions/#transitioncomponent-prop).
+   */
+  TransitionComponent?: React.ComponentType<TransitionProps>;
+}
+
+export interface PickerPopperProps extends ExportedPickerPopperProps, PaperProps {
+  role: 'tooltip' | 'dialog';
+  TrapFocusProps?: Partial<TrapFocusProps>;
+  anchorEl: PopperProps['anchorEl'];
+  open: PopperProps['open'];
+  onClose: () => void;
+  onOpen: () => void;
+}
+
+export const useStyles = makeStyles(
+  theme => ({
+    popper: {
+      zIndex: theme.zIndex.modal,
+    },
+    paper: {
+      transformOrigin: 'top center',
+      '&:focus': {
+        outline: 'auto',
+        [IS_TOUCH_DEVICE_MEDIA]: {
+          outline: 0,
+        },
+      },
+    },
+    topTransition: {
+      transformOrigin: 'bottom center',
+    },
+  }),
+  { name: 'MuiPickersPopper' }
+);
+
+export const PickerPopper: React.FC<PickerPopperProps> = ({
+  role,
+  PopperProps,
+  TrapFocusProps,
+  TransitionComponent = Grow,
+  open,
+  innerRef = null,
+  anchorEl,
+  children,
+  onClose,
+  onOpen,
+  ...other
+}) => {
+  const classes = useStyles();
+  const lastFocusedElementRef = React.useRef<Element | null>(null);
+  const paperRef = React.useRef<HTMLElement>(null);
+  const handlePopperRef = useForkRef(paperRef, innerRef);
+
+  useGlobalKeyDown(open, {
+    [keycode.Esc]: onClose,
+  });
+
+  React.useEffect(() => {
+    if (role === 'tooltip') {
+      return;
+    }
+
+    if (open) {
+      lastFocusedElementRef.current = document.activeElement;
+    } else if (
+      lastFocusedElementRef.current &&
+      lastFocusedElementRef.current instanceof HTMLElement
+    ) {
+      lastFocusedElementRef.current.focus();
+    }
+  }, [open, role]);
+
+  const handleBlur = () => {
+    if (!open) {
+      return;
+    }
+
+    // document.activeElement is updating on the next tick after `blur` called
+    executeInTheNextEventLoopTick(() => {
+      if (paperRef.current?.contains(document.activeElement)) {
+        return;
+      }
+
+      onClose();
+    });
+  };
+
+  return (
+    <Popper
+      transition
+      role={role}
+      open={open}
+      anchorEl={anchorEl}
+      placement="bottom"
+      className={clsx(classes.popper, PopperProps?.className)}
+      {...PopperProps}
+    >
+      {({ TransitionProps, placement }) => (
+        <TrapFocus
+          open={open}
+          disableAutoFocus
+          disableEnforceFocus={role === 'tooltip'}
+          isEnabled={() => true}
+          getDoc={() => paperRef.current?.ownerDocument ?? document}
```

Comment:
I would expect it to the always defined. If not, it might hide something else

```suggestion
          getDoc={() => paperRef.current.ownerDocument}
```

Reply:
https://github.com/mui-org/material-ui/blob/e0bd7326d160ac7757597786717eba77f986f2e4/packages/material-ui/src/utils/ownerDocument.js#L4

Reply:
Moreover document if optional field in typings, so there are some cases when it is undefined. And we must handle them

Reply:
The defensive logic for ownerDocument is meant to handle the case where the argument is already the document. In our case, it's a nested element in the page, it serves no purpose.

Regarding handling the optional field, `// @ts-ignore` sounds like the appropriate solution here, however, it looks like `getDoc` is call in an unsafe manner. It seems that we have to account for it here anyway. I'm resolving the comment;

Reply:
I cannot agree on such unsafe way. If typescript says that there is a possible case where `ownerDocument`, `ref.current` can be undefined – we **must** handle this case. 

This is, I suppose, the eternal problem of the javascript ecosystem. But this is how sound systems are built, and we as a core thing for a lot of projects cannot crash because of `cannot read property X of undefined`. We just cannot

Reply:
TypeScript can be as smart as the information he gets. He has no knowledge of the runtime guarantees. In this case. `ref.current` being defined is a constraint React enforce if accessed from an effect or layout effect. I think that the correct approach is `// @ts-ignore`, cc @eps1lon to get a broader point of view.

This is no different to what we explored in [2017 with flow](https://github.com/mui-org/material-ui/pull/9453). I think that we learned that it was better not to use defensive logic when we know the guarantee. If done otherwise, at best we add dead logic, at worse, we hide root issues.

Regarding `.ownerDocument` being defined is a constraint, if we look at the definitions, it's [null for a document](https://github.com/microsoft/TypeScript/blob/cc7cb9eadb10a56687513d4fc3a67d5210b387bb/src/lib/dom.generated.d.ts#L4546) and [defined for an element](https://github.com/microsoft/TypeScript/blob/cc7cb9eadb10a56687513d4fc3a67d5210b387bb/src/lib/dom.generated.d.ts#L5108). In our case, we always deal with an element, if will always be defined (if the element exists in the first place).

Reply:
I think it is a problem of developers who never worked with languages with a really strong type system (Rust for example). 

React guarantee that refs are filled, but it is not always true. In case when document is mounted by timeout, or unmounted from Dom manually. 

If you are consciously do not guard from such cases — the final product can crash in the most unexpected moment.